app/services/change_schedule.rb

```diff
@@ -131,6 +131,71 @@ def update_ecf_records!
+  end
+
+  def in_target_schedule?(induction_record)
+    induction_record.schedule == new_schedule
+  end
+
+  def historical_target_school_cohort(historical_record)
+    historical_record.school.school_cohorts.for_year(new_schedule.cohort_start_year).first.presence || SchoolCohort.create!(school: historical_record.school,
+                                                                                                                            cohort: new_schedule.cohort,
+                                                                                                                            induction_programme_choice: historical_record.induction_programme&.training_programme)
+  end
+
+  def historical_school_cohort_partnership(historical_record)
+    Partnership.create_with(relationship: true)
+    .find_or_create_by!(school: historical_record.school,
+                        cohort: new_schedule.cohort,
+                        lead_provider: historical_record.lead_provider,
+                        delivery_partner: historical_record.delivery_partner)
+  end
+
+  def historical_induction_programme(historical_record)
+    return historical_record.induction_programme if in_target_cohort?(historical_record)
+
+    historical_school_cohort = historical_target_school_cohort(historical_record)
+
+    set_default_programme = !historical_record.enrolled_in_fip? || !historical_record.lead_provider_id || !historical_record.delivery_partner_id
+
+    if set_default_programme
+      historical_school_cohort.default_induction_programme.presence || InductionProgramme.create_with(partnership: historical_record.partnership)
+                                                                                          .find_or_create_by!(school_cohort: historical_school_cohort,
+                                                                                                              training_programme: historical_record.induction_programme.training_programme)
+    else
+      historical_partnership = historical_school_cohort_partnership(historical_record)
+      historical_partnership.induction_programmes.first.presence || InductionProgramme
+                                                                    .full_induction_programme
+                                                                    .find_or_create_by!(school_cohort: historical_school_cohort, partnership: historical_partnership)
+    end
+  end
+
+  def update_historical_induction_records!
+    historical_records.all? do |historical_record|
+      next true if in_target?(historical_record)
+
+      begin
+        historical_record.update!(induction_programme: historical_induction_programme(historical_record),
+                                  schedule: new_schedule)
+      rescue ActiveRecord::RecordInvalid
+        false
+      end
+    end
+  end
+
+  def historical_records
+    return [] unless participant_profile
```

Comment:
should we return early if the profile is ECF? Also should we return early unless there is a cohort change required and not just schedule? Otherwise we will be changing schedule on all induction records unnecessarily